File: ios/HybridRiveFile.swift
import NitroModules
import RiveRuntime
#if RIVE_EXPERIMENTAL_API
@_spi(RiveExperimental) import RiveRuntime
#endif

typealias ReferencedAssetCache = [String: RiveFileAsset]

/// Source for creating experimental File instances
enum ExperimentalFileSource {
  case data(Data)
  case resource(String)
}

class HybridRiveFile: HybridRiveFileSpec, RiveViewSource {
  var riveFile: RiveFile?
  var referencedAssetCache: ReferencedAssetCache?
  var assetLoader: ReferencedAssetLoader?
  var cachedFactory: RiveFactory?
  private var weakViews: [Weak<RiveReactNativeView>] = []

  /// Source for experimental API - stored to create experimental File on demand
  var experimentalSource: ExperimentalFileSource?

  public func setRiveFile(_ riveFile: RiveFile) {
    self.riveFile = riveFile
  }

  func registerView(_ view: RiveReactNativeView) {
    weakViews.append(Weak(view))
  }

  func unregisterView(_ view: RiveReactNativeView) {
    weakViews.removeAll { $0.value === view }
  }

  private func refreshAfterAssetChange() {
    weakViews = weakViews.filter { $0.value != nil }

    for weakView in weakViews {
      guard let view = weakView.value else { continue }
      view.refreshAfterAssetChange()
    }
  }
  
  var viewModelCount: Double? {
    guard let count = riveFile?.viewModelCount else { return nil }
    return Double(count)
  }
  
  func viewModelByIndex(index: Double) throws -> (any HybridViewModelSpec)? {
    guard let vm = riveFile?.viewModel(at: UInt(index)) else { return nil }
    return HybridViewModel(viewModel: vm)
  }
  
  func viewModelByName(name: String) throws -> (any HybridViewModelSpec)? {
    guard let vm = riveFile?.viewModelNamed(name) else { return nil }
    return HybridViewModel(viewModel: vm)
  }
  
  func defaultArtboardViewModel(artboardBy: ArtboardBy?) throws -> (any HybridViewModelSpec)? {
    let artboard: RiveArtboard?
    
    if let artboardBy = artboardBy {
      switch artboardBy.type {
      case .index:
        guard let index = artboardBy.index else { return nil }
        artboard = try? riveFile?.artboard(from: Int(index))
      case .name:
        guard let name = artboardBy.name else { return nil }
        artboard = try? riveFile?.artboard(fromName: name)
      default:
        artboard = nil
      }
    } else {
      artboard = try? riveFile?.artboard()
    }
    
    guard let artboard = artboard,
          let vm = riveFile?.defaultViewModel(for: artboard) else { return nil }
    return HybridViewModel(viewModel: vm)
  }
  
  var artboardCount: Double {
    Double(riveFile?.artboardNames().count ?? 0)
  }

  var artboardNames: [String] {
    riveFile?.artboardNames() ?? []
  }

  func getBindableArtboard(name: String) throws -> any HybridBindableArtboardSpec {
    guard let bindable = try riveFile?.bindableArtboard(withName: name) else {
      throw NSError(
        domain: "RiveError",
        code: 0,
        userInfo: [NSLocalizedDescriptionKey: "Artboard '\(name)' not found"]
      )
    }
    return HybridBindableArtboard(bindableArtboard: bindable)
  }

  func updateReferencedAssets(referencedAssets: ReferencedAssetsType) {
    guard let assetsData = referencedAssets.data,
          let cache = referencedAssetCache,
          let loader = assetLoader,
          let _ = riveFile else {
      return
    }

    let dispatchGroup = DispatchGroup()
    var hasChanged = false

    for (key, assetData) in assetsData {
      guard let asset = cache[key] else { continue }
      if let riveFactory = cachedFactory {
        dispatchGroup.enter()
        loader.loadAsset(source: assetData, asset: asset, factory: riveFactory) {
          dispatchGroup.leave()
        }
      } else {
        RCTLogError("[RiveFile] no factory available for update")
      }
      hasChanged = true
    }

    if hasChanged {
      dispatchGroup.notify(queue: .main) { [weak self] in
        self?.refreshAfterAssetChange()
      }
    }
  }
  
  func getEnums() throws -> Promise<[RiveEnumDefinition]> {
    return Promise.async { [weak self] in
      #if RIVE_EXPERIMENTAL_API
      guard let source = self?.experimentalSource else {
        throw NSError(
          domain: "RiveError",
          code: 1,
          userInfo: [NSLocalizedDescriptionKey: "getEnums requires experimental API. Use USE_RIVE_SPM=1 with pod install."]
        )
      }

      // Create worker and experimental file on demand
      let worker = await Worker()
      let experimentalSource: Source
      switch source {
      case .data(let data):
        experimentalSource = .data(data)
      case .resource(let name):
        experimentalSource = .local(name, nil)
      }

      let file = try await File(source: experimentalSource, worker: worker)
      let viewModelEnums = try await file.getViewModelEnums()
      return viewModelEnums.map { vmEnum in
        RiveEnumDefinition(name: vmEnum.name, values: vmEnum.values)
      }
      #else
      throw NSError(
        domain: "RiveError",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "getEnums requires RiveRuntime 6.15.0+ with experimental API. Use USE_RIVE_SPM=1 with pod install."]
      )
      #endif
    }
  }

  func dispose() {
    weakViews.removeAll()
    referencedAssetCache = nil
    assetLoader = nil
    cachedFactory = nil
    riveFile = nil
  }

  deinit {
    dispose()
  }
}

File: android/src/legacy/java/com/margelo/nitro/rive/HybridRiveFile.kt
package com.margelo.nitro.rive

import androidx.annotation.Keep
import app.rive.runtime.kotlin.core.File
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.Promise
import java.lang.ref.WeakReference
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

@Keep
@DoNotStrip
class HybridRiveFile : HybridRiveFileSpec() {
  var riveFile: File? = null
  var referencedAssetCache: ReferencedAssetCache? = null
  var assetLoader: ReferencedAssetLoader? = null
  private val weakViews = mutableListOf<WeakReference<HybridRiveView>>()
  private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

  // Deprecated: Use getViewModelNamesAsync instead
  override val viewModelCount: Double?
    get() = riveFile?.viewModelCount?.toDouble()

  override fun getViewModelNamesAsync(): Promise<Array<String>> {
    return Promise.async {
      val count = riveFile?.viewModelCount ?: return@async emptyArray()
      (0 until count).mapNotNull { riveFile?.getViewModelByIndex(it)?.name }.toTypedArray()
    }
  }

  // Deprecated: Use getViewModelNamesAsync + viewModelByNameAsync instead
  override fun viewModelByIndex(index: Double): HybridViewModelSpec? {
    if (index < 0) return null
    return try {
      val vm = riveFile?.getViewModelByIndex(index.toInt()) ?: return null
      HybridViewModel(vm)
    } catch (e: Exception) {
      null
    }
  }

  // Deprecated: Use viewModelByNameAsync instead
  override fun viewModelByName(name: String): HybridViewModelSpec? {
    return try {
      val vm = riveFile?.getViewModelByName(name) ?: return null
      HybridViewModel(vm)
    } catch (e: Exception) {
      null
    }
  }

  // validate is ignored on legacy backend — native getViewModelByName(name) already returns null for unknown names
  override fun viewModelByNameAsync(name: String, validate: Boolean?): Promise<HybridViewModelSpec?> {
    return Promise.async {
      val vm = riveFile?.getViewModelByName(name) ?: return@async null
      HybridViewModel(vm)
    }
  }

  // Deprecated: Use defaultArtboardViewModelAsync instead
  override fun defaultArtboardViewModel(artboardBy: ArtboardBy?): HybridViewModelSpec? {
    try {
      val artboard = when (artboardBy?.type) {
        ArtboardByTypes.INDEX -> riveFile?.artboard(artboardBy.index!!.toInt())
        ArtboardByTypes.NAME -> riveFile?.artboard(artboardBy.name!!)
        null -> riveFile?.firstArtboard
      } ?: return null

      val vm = riveFile?.defaultViewModelForArtboard(artboard) ?: return null
      return HybridViewModel(vm)
    } catch (e: Exception) {
      return null
    }
  }

  override fun defaultArtboardViewModelAsync(artboardBy: ArtboardBy?): Promise<HybridViewModelSpec?> {
    return Promise.async { defaultArtboardViewModel(artboardBy) }
  }

  // Deprecated: Use getArtboardCountAsync instead
  override val artboardCount: Double
    get() = riveFile?.artboardNames?.size?.toDouble() ?: 0.0

  override fun getArtboardCountAsync(): Promise<Double> {
    return Promise.async {
      riveFile?.artboardNames?.size?.toDouble() ?: 0.0
    }
  }

  // Deprecated: Use getArtboardNamesAsync instead
  override val artboardNames: Array<String>
    get() = riveFile?.artboardNames?.toTypedArray() ?: emptyArray()

  override fun getArtboardNamesAsync(): Promise<Array<String>> {
    return Promise.async {
      riveFile?.artboardNames?.toTypedArray() ?: emptyArray()
    }
  }

  override fun getBindableArtboard(name: String): HybridBindableArtboardSpec {
    val file = riveFile ?: throw IllegalStateException("RiveFile not loaded")
    val bindable = file.createBindableArtboardByName(name)
    return HybridBindableArtboard(bindable)
  }

  fun registerView(view: HybridRiveView) {
    weakViews.add(WeakReference(view))
  }

  fun unregisterView(view: HybridRiveView) {
    weakViews.removeAll { it.get() == view }
  }

  private fun refreshAfterAssetChange() {
    weakViews.removeAll { it.get() == null }

    for (weakView in weakViews) {
      weakView.get()?.refreshAfterAssetChange()
    }
  }

  override fun updateReferencedAssets(referencedAssets: ReferencedAssetsType) {
    val assetsData = referencedAssets.data ?: return
    val cache = referencedAssetCache ?: return
    val loader = assetLoader ?: return

    val loadJobs = mutableListOf<kotlinx.coroutines.Deferred<Unit>>()

    for ((key, assetData) in assetsData) {
      val asset = cache[key] ?: continue
      loadJobs.add(loader.updateAsset(assetData, asset))
    }

    if (loadJobs.isNotEmpty()) {
      scope.launch {
        loadJobs.awaitAll()
        refreshAfterAssetChange()
      }
    }
  }

  override fun getEnums(): Promise<Array<RiveEnumDefinition>> {
    val file = riveFile ?: return Promise.resolved(emptyArray())
    return Promise.async {
      try {
        file.enums
          .map { enum ->
            RiveEnumDefinition(
              name = enum.name,
              values = enum.values.toTypedArray()
            )
          }.toTypedArray()
      } catch (e: NoSuchMethodError) {
        throw UnsupportedOperationException("getEnums requires rive-android SDK with enums support")
      }
    }
  }

  override fun dispose() {
    scope.cancel()
    weakViews.clear()
    assetLoader?.dispose()
    assetLoader = null
    riveFile?.release()
    riveFile = null
    referencedAssetCache?.clear()
    referencedAssetCache = null
  }
}
